### PR TITLE
[WIP][python worker] support set `--load-code-mode=local-only`

### PR DESCRIPTION
**Pull request by SongGuyang**
## Why are these changes needed?
- For some consideration about security in our case, we don't allow execute tasks by dynamic pickled function from a distrusted caller side. So we want to support this flag.
- Use like this: `ray start ---load-code-mode=local-only`

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by SongGuyang** [comment]
I need to add some tests.

**Comment by fyrestone** [comment]
There are some ways to load Python code:
- Load code from local
- Load code from pickled function / class
- Mixed above ones.

If ray is going to support all the modes, then the `--force-load-code-from-local` is not enough, maybe a `--load-code-mode` is better?

**Comment by fyrestone** [comment]
There are some ways to load Python code:
- Load code from local
- Load code from pickled function / class
- Mixed above ones.

If ray is going to support all the modes, then the `--force-load-code-from-local` is not enough, maybe a `--load-code-mode` is better?

**Comment by SongGuyang** [comment]
> There are some ways to load Python code:
> 
> * Load code from local
> * Load code from pickled function / class
> * Mixed above ones.
> 
> If ray is going to support all the modes, then the `--force-load-code-from-local` is not enough, maybe a `--load-code-mode` is better?

Sounds good.

**Comment by raulchen** [comment]
`--load-code-mode` makes more senses to me. Possible options will be "dynamic-distribution-only", "local-only", and "hybrid".  This will require a major user-api change. Let's create an issue with a formal proposal.

File: python/ray/exceptions.py [hunk 1]
```diff
@@ -204,6 +204,14 @@ def __str__(self):
                 "Check python-core-worker-*.log files for more information.")
 
 
+class FunctionLoadingError(RayError):
+    """Indicates that the function or class load failed."""
+
+    def __str__(self):
+        return ("Failed to load function or class from local. And this node "
+                "don't allow execute task from dynamic pickled function.")
```
Comment (MissiontoMars) [inline python/ray/exceptions.py:212]: Does each node use a different code loading strategy? from local or not from local.

File: python/ray/scripts/scripts.py [hunk 2]
```diff
@@ -488,8 +494,8 @@ def start(node_ip_address, address, port, redis_password, redis_shard_ports,
           autoscaling_config, no_redirect_worker_output, no_redirect_output,
           plasma_store_socket_name, raylet_socket_name, temp_dir,
           system_config, enable_object_reconstruction, metrics_export_port,
-          no_monitor, tracing_startup_hook, ray_debugger_external, log_style,
-          log_color, verbose):
+          no_monitor, tracing_startup_hook, ray_debugger_external,
+          force_load_code_from_local, log_style, log_color, verbose):
```
Comment (MissiontoMars) [inline python/ray/scripts/scripts.py:498]: Better to append `force_load_code_from_local` to the last parameter.

File: python/ray/workers/default_worker.py [hunk 3]
```diff
@@ -134,6 +134,11 @@
     default=False,
     action="store_true",
     help="True if Ray debugger is made available externally.")
+parser.add_argument(
+    "--force-load-code-from-local",
+    default=False,
+    action="store_true",
+    help="If true, don't allow execute task by dynamic pickled function.")
```
Comment (MissiontoMars) [inline python/ray/workers/default_worker.py:141]: Is `Executing task by dynamic pickled function is not allowed.` sounds a little better?

File: src/ray/protobuf/common.proto [hunk 4]
```diff
@@ -116,40 +116,42 @@ enum ErrorType {
   // but this object is not reconstructable (created by actor, ray.put, or a
   // borowed ObjectRef).
   OBJECT_UNRECONSTRUCTABLE = 2;
+  // Indicates that a task failed because the function or class couldn't be loaded.
+  FUNCTION_LOADING_ERROR = 3;
```
Comment (MissiontoMars) [inline src/ray/protobuf/common.proto:120]: Could we append `FUNCTION_LOADING_ERROR` to the last one, 15? The change here cause the following invalid modifications.

File: python/ray/workers/default_worker.py [hunk 5]
```diff
@@ -134,6 +134,11 @@
     default=False,
     action="store_true",
     help="True if Ray debugger is made available externally.")
+parser.add_argument(
```
Comment (ericl) [inline python/ray/workers/default_worker.py:137]: Can we make this a private/experimental API for now?

File: src/ray/core_worker/transport/direct_task_transport.cc [hunk 6]
```diff
@@ -644,70 +644,72 @@ void CoreWorkerDirectTaskSubmitter::PushNormalTask(
...
+    if (!status.ok()) {
+      // TODO: It'd be nice to differentiate here between process vs node
+      // failure (e.g., by contacting the raylet). If it was a process
+      // failure, it may have been an application-level error and it may
+      // not make sense to retry the task.
+      rpc::ErrorType error_type =
+          is_actor ? rpc::ErrorType::ACTOR_DIED : rpc::ErrorType::WORKER_DIED;
+      if (status.IsFunctionLoadingError()) {
+        error_type = rpc::ErrorType::FUNCTION_LOADING_ERROR;
+      }
```
Comment (ericl) [inline src/ray/core_worker/transport/direct_task_transport.cc:703]: Is it necessary to add FUNCTION_LOADING_ERROR? Could we raise an application-level error instead (like raise a python exception?)

**Review by ericl — COMMENTED** [review]
I'm a bit confused how Python works at all with this disabled. Aren't all functions defined dynamically?

File: python/ray/workers/default_worker.py (hunk 5 above)
Comment (SongGuyang) [reply python/ray/workers/default_worker.py:137]: I set `hidden=True`, is it enough?

File: src/ray/core_worker/transport/direct_task_transport.cc (hunk 6 above)
Comment (SongGuyang) [reply src/ray/core_worker/transport/direct_task_transport.cc:703]: I had add a new exception type `FunctionLoadingError` in python. And the error type `rpc::ErrorType::FUNCTION_LOADING_ERROR` is used to tell the caller to throw the python level exception. Do we have an easier way to do this?

**Comment by SongGuyang** [comment]
> I'm a bit confused how Python works at all with this disabled. Aren't all functions defined dynamically?

If we set `--load-code-mode=local-only`, the python job must set the `code_search_path` to make sure the remote function could be loaded in local file system.

File: python/ray/exceptions.py (hunk 1 above)
Comment (SongGuyang) [reply python/ray/exceptions.py:212]: Yep, we can allow different code load mode in one cluster.

File: python/ray/scripts/scripts.py (hunk 2 above)
Comment (SongGuyang) [reply python/ray/scripts/scripts.py]: I can't do this because the last three params are built-in.

**Comment by wuisawesome** [comment]
> For some consideration about security in our case, we don't allow execute tasks by dynamic pickled function from a distrusted caller side. So we want to support this flag.

Actually, I have a question here. If we don't want to dynamically pickle functions, wouldn't it be easier to just use pickle instead of cloudpickle? Regular pickle to points to functions, which is why cloudpickle exists to begin with.

**Comment by fyrestone** [comment]
> > For some consideration about security in our case, we don't allow execute tasks by dynamic pickled function from a distrusted caller side. So we want to support this flag.
> 
> Actually, I have a question here. If we don't want to dynamically pickle functions, wouldn't it be easier to just use pickle instead of cloudpickle? Regular pickle to points to functions, which is why cloudpickle exists to begin with.

There are many useful features that cloudpickle supports, e.g. pickling `memoryview` object. I think we can still use cloudpickle to keep ray version compatibility. If you have security concern, the `pickle` is not secure, too. please refer to: https://docs.python.org/3/library/pickle.html

```
Warning The pickle module is not secure. Only unpickle data you trust.
```

**Comment by wuisawesome** [comment]
yeah i think we discussed this in the meeting last night, but I thought we agreed that this PR doesn't provide any security since we're still pickling objects, and an attacker can cause arbitrary code execution via pickling an object (whether we load code doesn't matter), so this PR is just making it a bit more obscure. 

btw cloudpickle doesn't actually have a custom depickler so the deserialization is identical between cloudpickle and pickle.

**Comment by SongGuyang** [comment]
@wuisawesome I think that are two different problems. Current PR only protects the nodes from executing unsafe code. We want to limit that all the remote function should be an existing function in callee side. And it is not related to object fetching security.

**Comment by SongGuyang** [comment]
@wuisawesome Sorry, I have misunderstood your comment. Anyway, do you think we should support current PR's feature?

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by kfstorm** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've updated our formatting configuration for C++ code. (see #22725)

This PR includes C++ code change. To prevent issues with merging your code, here's what you'll need to do:

1. Merge the latest changes from upstream/master branch into your branch.

```bash
git pull upstream master
git merge upstream/master
```

2. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated C++ formatting configuration.

3. Format changed files.

```bash
scripts/format.sh
```

4. Commit your changes.

```bash
git add --all
git commit -m "Format C++ code"
```

**Pull request closed without merging**